### PR TITLE
[PIPE-4937] update requires docs to allow dependency on different statuses

### PR DESCRIPTION
**Pull request by SashaLsnko**
# Description
Update documentation of `requires` job field. 

# Reasons
[Ticket.](https://circleci.atlassian.net/browse/PIPE-4937)
The new functionality allows users to define what statuses the upstream job dependency should have for the dependent job to start running.

# Content Checklist
Please follow our style when contributing to CircleCI docs. Our style guide is here: [https://circleci.com/docs/style/style-guide-overview](https://circleci.com/docs/style/style-guide-overview).

Please take a moment to check through the following items when submitting your PR (this is just a guide so will not be relevant for all PRs) 😸:

- [ ] Break up walls of text by adding paragraph breaks.
- [ ] Consider if the content could benefit from more structure, such as lists or tables, to make it easier to consume.
- [ ] Keep the title between 20 and 70 characters.
- [ ] Consider whether the content would benefit from more subsections (h2-h6 headings) to make it easier to consume.
- [ ] Check all headings h1-h6 are in sentence case (only first letter is capitalized).
- [ ] Is there a "Next steps" section at the end of the page giving the reader a clear path to what to read next?
- [ ] Include relevant backlinks to other CircleCI docs/pages.

### CODE REVIEW

File: jekyll/_cci2/configuration-reference.adoc [hunk 1]
```diff
@@ -2628,9 +2628,40 @@ Jobs are run concurrently by default, so you must explicitly require any depende
 | `requires`
 | N
 | List
-| A list of jobs that must succeed for the job to start. Note: When jobs in the current workflow that are listed as dependencies are not executed (due to a filter function for example), their requirement as a dependency for other jobs will be ignored by the requires option. However, if all dependencies of a job are filtered, then that job will not be executed either.
+| A list of jobs that must succeed or attain a certain status for the job to start. Note: When jobs in the current workflow that are listed as dependencies are not executed (due to a filter function for example), their requirement as a dependency for other jobs will be ignored by the requires option. However, if all dependencies of a job are filtered, then that job will not be executed either.
```
Comment (briceicle) [inline jekyll/_cci2/configuration-reference.adoc:2631]: ~ I would rephrase to "A list of jobs that must succeed or attain a specified status..."

File: jekyll/_cci2/configuration-reference.adoc [hunk 2]
```diff
@@ -2628,9 +2628,40 @@ Jobs are run concurrently by default, so you must explicitly require any depende
...
+|===
+| Key | Type | Description
+
+| `job_name`
+| Job Name
+| An upstream job that must `succeed` for the job to start.
+
+| `job_name`
+| Map
+| A map of job name to a list of statuses (possible values are `success`, `failed` and `canceled`). The job starts if the upstream dependency reaches one of the statuses.
```
Comment (briceicle) [inline jekyll/_cci2/configuration-reference.adoc:2645]: ~ I suggest putting the possible values in a separate sentence
```
A map of required job to a list of statuses. 
The job starts if the required job reaches one of the statuses. The possible status values are: `success`, `failed` and `canceled`.
```

File: jekyll/_cci2/configuration-reference.adoc [hunk 3]
```diff
@@ -2628,9 +2628,40 @@ Jobs are run concurrently by default, so you must explicitly require any depende
...
 |===
 
+[.table.table-striped]
+[cols=3*, options="header", stripes=even]
+|===
+| Key | Type | Description
+
+| `job_name`
+| Job Name
+| An upstream job that must `succeed` for the job to start.
```
Comment (briceicle) [inline jekyll/_cci2/configuration-reference.adoc:2641]: I'd rename this to `required job` to be consistent with what customers write in their config. 
"upstream" is a term we use internally and is not mentioned anywhere else in our docs.

File: jekyll/_cci2/configuration-reference.adoc [hunk 4]
```diff
@@ -2628,9 +2628,40 @@ Jobs are run concurrently by default, so you must explicitly require any depende
...
+  my-workflow:
+    jobs:
+      - build
+      - test:
+          requires:
+            - build
+      - deploy:
+          requires:
+            - build
+            - test:
```
Comment (briceicle) [inline jekyll/_cci2/configuration-reference.adoc:2660]: I suggest adding an example to show it is possible to define a single status on the same line as the required job name, e.g.
```
requires:
  - foo: canceled
```

**Review by briceicle — APPROVED** [review]

**Comment by SashaLsnko** [comment]
I addressed your comments @briceicle

**Review by briceicle — APPROVED** [review]

File: jekyll/_cci2/configuration-reference.adoc [hunk 5]
```diff
@@ -2628,9 +2628,46 @@ Jobs are run concurrently by default, so you must explicitly require any depende
...
+            - build
+      - notify-canceled:
+          requires:
+            - build: canceled
+      - deploy:
+          requires:
+            - build
+            - test:
+              - success
+              - failed
```
Comment (rosieyohannan) [inline jekyll/_cci2/configuration-reference.adoc:2667]: Does it make sense to have an example where you deploy on failed or passed tests, just not cancelled?

Comment (SashaLsnko) [reply jekyll/_cci2/configuration-reference.adoc:2667]: I can rename the `deploy` job to be `share-test-results` for example. Would that make more sense?

Comment (briceicle) [reply jekyll/_cci2/configuration-reference.adoc:2667]: If we want a realistic example, deploying on failed isn't really a use case people typically want. A common use case for this feature is to run a clean up job when a deploy fails to tear down infrastructure/resources.

So we could have a `deploy` job that requires `build` and `test`. Then a `cleanup` job that requires `deploy` to fail.

For example:
```yaml
workflows:
  my-workflow:
    jobs:
      - build
      - test:
          requires:
            - build
      - deploy:
          requires:
            - build
            - test
     - notify-slack:
          requires:
            - build: canceled
      - cleanup:
          requires:
            - deploy:
              - failed
              - canceled
```

**Review by rosieyohannan — APPROVED** [review]
